File: sdks/python/golem_sdk/types/option.py
# Copyright 2024-2025 Golem Cloud
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Option type for representing optional values."""

from typing import Callable, Generic, Optional, TypeVar, Union, overload

T = TypeVar("T")
U = TypeVar("U")


class Option(Generic[T]):
    """
    Represents an optional value that may or may not be present.

    This is similar to Rust's Option type or Haskell's Maybe type.

    Example:
        ```python
        value: Option[int] = Option.some(42)
        empty: Option[int] = Option.none()

        # Check if value is present
        if value.is_some():
            print(value.unwrap())  # 42

        # Map over the value
        doubled = value.map(lambda x: x * 2)  # Option.some(84)

        # Get with default
        result = empty.unwrap_or(0)  # 0
        ```
    """

    __slots__ = ("_value", "_is_some")

    def __init__(self, value: Optional[T], is_some: bool) -> None:
        self._value = value
        self._is_some = is_some

    @classmethod
    def some(cls, value: T) -> "Option[T]":
        """Create an Option containing a value."""
        return cls(value, True)

    @classmethod
    def none(cls) -> "Option[T]":
        """Create an empty Option."""
        return cls(None, False)

    @classmethod
    def from_optional(cls, value: Optional[T]) -> "Option[T]":
        """Create an Option from a Python Optional."""
        if value is None:
            return cls.none()
        return cls.some(value)

    def is_some(self) -> bool:
        """Check if the Option contains a value."""
        return self._is_some

    def is_none(self) -> bool:
        """Check if the Option is empty."""
        return not self._is_some

    def unwrap(self) -> T:
        """
        Get the contained value.

        Raises:
            ValueError: If the Option is empty.
        """
        if not self._is_some:
            raise ValueError("Called unwrap on a None Option")
        return self._value  # type: ignore

    def unwrap_or(self, default: T) -> T:
        """Get the contained value or a default."""
        if self._is_some:
            return self._value  # type: ignore
        return default

    def unwrap_or_else(self, f: Callable[[], T]) -> T:
        """Get the contained value or compute a default."""
        if self._is_some:
            return self._value  # type: ignore
        return f()

    def map(self, f: Callable[[T], U]) -> "Option[U]":
        """Transform the contained value if present."""
        if self._is_some:
            return Option.some(f(self._value))  # type: ignore
        return Option.none()

    def flat_map(self, f: Callable[[T], "Option[U]"]) -> "Option[U]":
        """Transform the contained value with a function that returns an Option."""
        if self._is_some:
            return f(self._value)  # type: ignore
        return Option.none()

    def filter(self, predicate: Callable[[T], bool]) -> "Option[T]":
        """Keep the value only if it satisfies the predicate."""
        if self._is_some and predicate(self._value):  # type: ignore
            return self
        return Option.none()

    def to_optional(self) -> Optional[T]:
        """Convert to a Python Optional."""
        if self._is_some:
            return self._value
        return None

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Option):
            return NotImplemented
        if self._is_some != other._is_some:
            return False
        if not self._is_some:
            return True
        return self._value == other._value

    def __repr__(self) -> str:
        if self._is_some:
            return f"Option.some({self._value!r})"
        return "Option.none()"

    def __hash__(self) -> int:
        if self._is_some:
            return hash(("some", self._value))
        return hash(("none",))


# Convenience functions
def some(value: T) -> Option[T]:
    """Create an Option containing a value."""
    return Option.some(value)


def none() -> Option[T]:
    """Create an empty Option."""
    return Option.none()

File: sdks/java/golem-sdk/src/main/java/cloud/golem/sdk/types/Option.java
/*
 * Copyright 2024-2025 Golem Cloud
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cloud.golem.sdk.types;

import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * Represents an optional value that may or may not be present.
 *
 * <p>This is similar to Java's Optional but provides additional methods
 * for compatibility with Golem's type system.
 *
 * @param <T> The type of the contained value
 */
public final class Option<T> {

    private final T value;
    private final boolean present;

    private Option(T value, boolean present) {
        this.value = value;
        this.present = present;
    }

    /**
     * Create an Option containing a value.
     *
     * @param value The value to contain
     * @param <T> The type of the value
     * @return An Option containing the value
     */
    public static <T> Option<T> some(T value) {
        return new Option<>(Objects.requireNonNull(value), true);
    }

    /**
     * Create an empty Option.
     *
     * @param <T> The type of the absent value
     * @return An empty Option
     */
    @SuppressWarnings("unchecked")
    public static <T> Option<T> none() {
        return new Option<>(null, false);
    }

    /**
     * Create an Option from a nullable value.
     *
     * @param value The value (may be null)
     * @param <T> The type of the value
     * @return Option.some(value) if non-null, Option.none() otherwise
     */
    public static <T> Option<T> fromNullable(T value) {
        return value != null ? some(value) : none();
    }

    /**
     * Create an Option from a Java Optional.
     *
     * @param optional The Java Optional
     * @param <T> The type of the value
     * @return An Option equivalent to the Optional
     */
    public static <T> Option<T> fromOptional(Optional<T> optional) {
        return optional.map(Option::some).orElseGet(Option::none);
    }

    /**
     * Check if the Option contains a value.
     *
     * @return true if a value is present
     */
    public boolean isSome() {
        return present;
    }

    /**
     * Check if the Option is empty.
     *
     * @return true if no value is present
     */
    public boolean isNone() {
        return !present;
    }

    /**
     * Get the contained value.
     *
     * @return The contained value
     * @throws NoSuchElementException If the Option is empty
     */
    public T unwrap() {
        if (!present) {
            throw new NoSuchElementException("Called unwrap on a None Option");
        }
        return value;
    }

    /**
     * Get the contained value or a default.
     *
     * @param defaultValue The default value
     * @return The contained value or the default
     */
    public T unwrapOr(T defaultValue) {
        return present ? value : defaultValue;
    }

    /**
     * Get the contained value or compute a default.
     *
     * @param supplier Supplier for the default value
     * @return The contained value or the computed default
     */
    public T unwrapOrElse(Supplier<T> supplier) {
        return present ? value : supplier.get();
    }

    /**
     * Transform the contained value if present.
     *
     * @param mapper The transformation function
     * @param <U> The result type
     * @return An Option containing the transformed value, or None
     */
    public <U> Option<U> map(Function<? super T, ? extends U> mapper) {
        if (present) {
            return some(mapper.apply(value));
        }
        return none();
    }

    /**
     * Transform the contained value with a function that returns an Option.
     *
     * @param mapper The transformation function
     * @param <U> The result type
     * @return The result of the transformation, or None
     */
    public <U> Option<U> flatMap(Function<? super T, Option<U>> mapper) {
        if (present) {
            return mapper.apply(value);
        }
        return none();
    }

    /**
     * Keep the value only if it satisfies the predicate.
     *
     * @param predicate The predicate to test
     * @return This Option if present and predicate is true, None otherwise
     */
    public Option<T> filter(Predicate<? super T> predicate) {
        if (present && predicate.test(value)) {
            return this;
        }
        return none();
    }

    /**
     * Execute an action if a value is present.
     *
     * @param action The action to execute
     */
    public void ifSome(Consumer<? super T> action) {
        if (present) {
            action.accept(value);
        }
    }

    /**
     * Convert to a Java Optional.
     *
     * @return An equivalent Java Optional
     */
    public Optional<T> toOptional() {
        return present ? Optional.of(value) : Optional.empty();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Option<?> option = (Option<?>) o;
        if (present != option.present) return false;
        return Objects.equals(value, option.value);
    }

    @Override
    public int hashCode() {
        return present ? Objects.hash(value) : 0;
    }

    @Override
    public String toString() {
        return present ? "Option.some(" + value + ")" : "Option.none()";
    }
}
